File: clinic/clinic/page/clinic_dashboard/clinic_dashboard.py
import frappe
from frappe import _
from frappe.utils import flt, getdate


@frappe.whitelist()
def get_dashboard_data(from_date=None, to_date=None, clinic=None, doctor=None, lead_source=None):
    """
    Single endpoint returning all dashboard metrics.
    Optimized: 4 queries instead of 7, with combined aggregations.
    """
    filters = {
        "from_date": getdate(from_date) if from_date else None,
        "to_date": getdate(to_date) if to_date else None,
        "clinic": clinic or None,
        "doctor": doctor or None,
        "lead_source": lead_source or None,
    }

    # ── Query 1: Appointments by status (also gives total) ──────────────
    appointments_by_status = _get_appointments_by_status(filters)
    total_appointments = sum(r.get("count", 0) for r in appointments_by_status)

    # ── Query 2: Revenue summary + per-clinic breakdown (single query) ──
    revenue_rows = _get_revenue_data(filters)
    revenue_summary = {
        "total_invoiced": 0, "total_paid": 0,
        "total_outstanding": 0, "total_discount": 0, "total_taxes": 0,
    }
    revenue_by_clinic = []
    for r in revenue_rows:
        revenue_summary["total_invoiced"] += flt(r.total_invoiced)
        revenue_summary["total_paid"] += flt(r.total_paid)
        revenue_summary["total_outstanding"] += flt(r.total_outstanding)
        revenue_summary["total_discount"] += flt(r.total_discount)
        revenue_summary["total_taxes"] += flt(r.total_taxes)
        revenue_by_clinic.append({
            "clinic": r.clinic,
            "total_invoiced": flt(r.total_invoiced),
            "total_paid": flt(r.total_paid),
        })

    # ── Query 3: Lead source + referrals (two lightweight queries) ──────
    lead_source_performance = _get_lead_source_performance(filters)
    referral_stats = _get_referral_stats(filters)

    # ── Query 4: Doctor financial (single query) ────────────────────────
    doctor_financial = _get_doctor_financial(filters)

    return {
        "appointment_summary": {"total_appointments": total_appointments},
        "appointments_by_status": appointments_by_status,
        "revenue_summary": revenue_summary,
        "revenue_by_clinic": revenue_by_clinic,
        "lead_source_performance": lead_source_performance,
        "referral_stats": referral_stats,
        "doctor_financial": doctor_financial,
    }


# ── helpers ──────────────────────────────────────────────────────────────────

def _appt_conds(filters, alias="ca"):
    cond = ""
    if filters.get("from_date"):
        cond += f" AND {alias}.appointment_date >= %(from_date)s"
    if filters.get("to_date"):
        cond += f" AND {alias}.appointment_date <= %(to_date)s"
    if filters.get("clinic"):
        cond += f" AND {alias}.clinic = %(clinic)s"
    if filters.get("doctor"):
        cond += f" AND {alias}.physician = %(doctor)s"
    return cond


def _inv_conds(filters, alias="SI"):
    cond = ""
    if filters.get("from_date"):
        cond += f" AND {alias}.posting_date >= %(from_date)s"
    if filters.get("to_date"):
        cond += f" AND {alias}.posting_date <= %(to_date)s"
    if filters.get("doctor"):
        cond += f" AND {alias}.doctor = %(doctor)s"
    return cond


# ── 1. Appointments by status (replaces two separate queries) ───────────────

def _get_appointments_by_status(filters):
    conds = _appt_conds(filters, "PA")
    return frappe.db.sql("""
        SELECT PA.status AS status, COUNT(*) AS count
        FROM `tabClient Appointment CT` PA
        WHERE PA.status != 'Cancelled' {conds}
        GROUP BY PA.status
        ORDER BY count DESC
    """.format(conds=conds), filters, as_dict=True)


# ── 2. Revenue data: per-clinic with totals derived in Python ───────────────

def _get_revenue_data(filters):
    conds = _inv_conds(filters)
    clinic_cond = " AND PA.clinic = %(clinic)s" if filters.get("clinic") else ""

    return frappe.db.sql("""
        SELECT
            IFNULL(PA.clinic, 'Unknown') AS clinic,
            COALESCE(SUM(SI.grand_total), 0) AS total_invoiced,
            COALESCE(SUM(SI.paid_amount), 0) AS total_paid,
            COALESCE(SUM(SI.outstanding_amount), 0) AS total_outstanding,
            COALESCE(SUM(SI.discount_amount), 0) AS total_discount,
            COALESCE(SUM(SI.total_taxes_and_charges), 0) AS total_taxes
        FROM `tabSales Invoice` SI
        INNER JOIN `tabClient Appointment CT` PA ON PA.name = SI.appointment
        WHERE SI.docstatus = 1 {conds} {clinic_cond}
        GROUP BY PA.clinic
        ORDER BY total_invoiced DESC
        LIMIT 15
    """.format(conds=conds, clinic_cond=clinic_cond), filters, as_dict=True)


# ── 3. Lead source (optimized: pre-materialized min-creation subquery) ──────

def _get_lead_source_performance(filters):
    if not filters.get("from_date") or not filters.get("to_date"):
        return []

    date_cond = ""
    if filters.get("from_date"):
        date_cond += " AND si.posting_date >= %(from_date)s"
    if filters.get("to_date"):
        date_cond += " AND si.posting_date <= %(to_date)s"
    lead_cond = " AND c.lead_source = %(lead_source)s" if filters.get("lead_source") else ""

    return frappe.db.sql("""
        SELECT
            IFNULL(c.lead_source, 'Unknown') AS lead_source,
            SUM(CASE WHEN mini.first_inv BETWEEN %(from_date)s AND %(to_date)s
                     THEN si.paid_amount ELSE 0 END) AS new_billing,
            SUM(CASE WHEN mini.first_inv < %(from_date)s
                     THEN si.paid_amount ELSE 0 END) AS repeat_billing
        FROM `tabCustomer` c
        INNER JOIN `tabSales Invoice` si
            ON c.name = si.customer AND si.docstatus = 1
            AND si.status IN ('Paid', 'Overdue', 'Partly Paid')
            {date_cond}
        INNER JOIN (
            SELECT customer, MIN(creation) AS first_inv
            FROM `tabSales Invoice`
            WHERE docstatus = 1
            GROUP BY customer
        ) mini ON mini.customer = c.name
        {lead_cond}
        GROUP BY c.lead_source
        HAVING new_billing > 0 OR repeat_billing > 0
        ORDER BY new_billing DESC
        LIMIT 15
    """.format(date_cond=date_cond, lead_cond=lead_cond), filters, as_dict=True)


# ── 4. Referral stats (single aggregate, no N+1) ───────────────────────────

def _get_referral_stats(filters):
    fwd_cond = ""
    if filters.get("from_date"):
        fwd_cond += " AND af.date >= %(from_date)s"
    if filters.get("to_date"):
        fwd_cond += " AND af.date <= %(to_date)s"
    if filters.get("doctor"):
        fwd_cond += " AND af.from_doctor = %(doctor)s"

    row = frappe.db.sql("""
        SELECT
            COUNT(*) AS total_referrals,
            SUM(af.bill_status = 'Billed') AS billed_referrals,
            SUM(IFNULL(af.bill_status, '') != 'Billed') AS non_billed_referrals
        FROM `tabAppointment Forward` af
        WHERE af.from_doctor != af.to_doctor
        AND af.docstatus < 2
        {fwd_cond}
    """.format(fwd_cond=fwd_cond), filters, as_dict=True)
    return row[0] if row else {"total_referrals": 0, "billed_referrals": 0, "non_billed_referrals": 0}


# ── 5. Doctor financial (single query, avg computed in SQL) ─────────────────

def _get_doctor_financial(filters):
    conds = _appt_conds(filters, "ca")
    return frappe.db.sql("""
        SELECT
            ca.physician AS doctor,
            d.first_name AS doctor_name,
            COUNT(DISTINCT ca.name) AS appointment_count,
            COALESCE(SUM(si.grand_total), 0) AS total_invoiced,
            COALESCE(SUM(si.grand_total - si.outstanding_amount), 0) AS total_paid,
            COALESCE(SUM(si.outstanding_amount), 0) AS total_outstanding,
            ROUND(COALESCE(SUM(si.grand_total - si.outstanding_amount), 0)
                  / GREATEST(COUNT(DISTINCT ca.name), 1), 2) AS average_per_appointment
        FROM `tabClient Appointment CT` ca
        INNER JOIN `tabDoctor` d ON ca.physician = d.name
        LEFT JOIN `tabSales Invoice` si
            ON si.appointment = ca.name AND si.docstatus = 1
        WHERE ca.status IN ('waiting attend', 'Attended')
        AND ca.docstatus < 2
        {conds}
        GROUP BY ca.physician, d.first_name
        ORDER BY total_paid DESC
        LIMIT 15
    """.format(conds=conds), filters, as_dict=True)

File: clinic/clinic/page/clinic_dashboard/clinic_dashboard.js
frappe.pages['clinic-dashboard'].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Manager Analytics Dashboard',
		single_column: true
	});

	// Attach the dashboard instance to the page so it persists
	page.dashboard = new ClinicDashboard(page);
};

class ClinicDashboard {
	constructor(page) {
		this.page = page;
		this.$parent = $(page.body);
		this.charts = {};
		this.init();
	}

	init() {
		this.render_skeleton();
		this.bind_events();
		this.fetch_data();
	}

	// ── Skeleton / layout ────────────────────────────────────────────────

	render_skeleton() {
		this.$parent.html(`
            <div class="clinic-dashboard-wrapper">
                <!-- Filters -->
                <div class="dashboard-filters" id="dash-filters">
                    <div class="filter-group">
                        <label>${__('From Date')}</label>
                        <input type="date" id="filter-from-date"
                            value="${frappe.datetime.month_start()}" />
                    </div>
                    <div class="filter-group">
                        <label>${__('To Date')}</label>
                        <input type="date" id="filter-to-date"
                            value="${frappe.datetime.month_end()}" />
                    </div>
                    <div class="filter-group">
                        <label>${__('Clinic')}</label>
                        <input type="text" id="filter-clinic"
                               placeholder="${__('All Clinics')}" />
                    </div>
                    <div class="filter-group">
                        <label>${__('Doctor')}</label>
                        <input type="text" id="filter-doctor"
                               placeholder="${__('All Doctors')}" />
                    </div>
                    <div class="filter-group">
                        <label>${__('Lead Source')}</label>
                        <input type="text" id="filter-lead-source"
                               placeholder="${__('All Sources')}" />
                    </div>
                    <button class="btn-refresh" id="btn-refresh">
                        <svg width="14" height="14" viewBox="0 0 24 24" fill="none"
                            stroke="currentColor" stroke-width="2.5"
                            stroke-linecap="round" stroke-linejoin="round">
                            <polyline points="23 4 23 10 17 10"/>
                            <path d="M20.49 15a9 9 0 1 1-2.12-9.36L23 10"/>
                        </svg>
                        ${__('Refresh')}
                    </button>
                </div>

                <!-- KPI row -->
                <div class="kpi-grid" id="kpi-grid"></div>

                <!-- Referral summary -->
                <div class="referral-grid" id="referral-grid"></div>

                <!-- Charts 2×2 -->
                <div class="charts-grid" id="charts-grid">
                    <div class="chart-card">
                        <div class="chart-title">
                            <span class="dot" style="background:var(--accent-blue)"></span>
                            ${__('Appointments by Status')}
                        </div>
                        <div id="chart-status"></div>
                    </div>
                    <div class="chart-card">
                        <div class="chart-title">
                            <span class="dot" style="background:var(--accent-green)"></span>
                            ${__('Revenue by Clinic')}
                        </div>
                        <div id="chart-clinic-revenue"></div>
                    </div>
                    <div class="chart-card">
                        <div class="chart-title">
                            <span class="dot" style="background:var(--accent-purple)"></span>
                            ${__('Lead Source Performance')}
                        </div>
                        <div id="chart-lead-source"></div>
                    </div>
                    <div class="chart-card">
                        <div class="chart-title">
                            <span class="dot" style="background:var(--accent-amber)"></span>
                            ${__('Top Doctors by Revenue')}
                        </div>
                        <div id="chart-top-doctors"></div>
                    </div>
                </div>

                <!-- Doctor financial table -->
                <div class="doctor-table-card" id="doctor-table-card">
                    <div class="table-title">${__('Doctor Financial Summary')}</div>
                    <table class="doctor-table" id="doctor-table">
                        <thead>
                            <tr>
                                <th>${__('Doctor')}</th>
                                <th>${__('Appointments')}</th>
                                <th>${__('Total Invoiced')}</th>
                                <th>${__('Total Paid')}</th>
                                <th>${__('Outstanding')}</th>
                                <th>${__('Avg / Appointment')}</th>
                            </tr>
                        </thead>
                        <tbody></tbody>
                    </table>
                </div>
            </div>
        `);

		// Make Link-style auto-completes for filter inputs
		this.setup_link_fields();
	}

	setup_link_fields() {
		const me = this;
		// Clinic → Link to Department
		this.clinic_control = frappe.ui.form.make_control({
			df: {
				fieldtype: 'Link',
				options: 'Department',
				fieldname: 'clinic',
				placeholder: __('All Clinics'),
				only_input: true,
			},
			parent: this.$parent.find('#filter-clinic').parent(),
			render_input: true,
		});
		this.clinic_control.$input.attr('id', 'filter-clinic-link');
		this.$parent.find('#filter-clinic').hide();

		// Doctor → Link to Doctor
		this.doctor_control = frappe.ui.form.make_control({
			df: {
				fieldtype: 'Link',
				options: 'Doctor',
				fieldname: 'doctor',
				placeholder: __('All Doctors'),
				only_input: true,
			},
			parent: this.$parent.find('#filter-doctor').parent(),
			render_input: true,
		});
		this.doctor_control.$input.attr('id', 'filter-doctor-link');
		this.$parent.find('#filter-doctor').hide();

		// Lead Source → Link to Lead Source
		this.lead_source_control = frappe.ui.form.make_control({
			df: {
				fieldtype: 'Link',
				options: 'Lead Source',
				fieldname: 'lead_source',
				placeholder: __('All Sources'),
				only_input: true,
			},
			parent: this.$parent.find('#filter-lead-source').parent(),
			render_input: true,
		});
		this.lead_source_control.$input.attr('id', 'filter-lead-source-link');
		this.$parent.find('#filter-lead-source').hide();
	}

	// ── Events ───────────────────────────────────────────────────────────

	bind_events() {
		this.$parent.find('#btn-refresh').on('click', () => this.fetch_data());
	}

	get_filters() {
		return {
			from_date: this.$parent.find('#filter-from-date').val(),
			to_date: this.$parent.find('#filter-to-date').val(),
			clinic: this.clinic_control ? this.clinic_control.get_value() : '',
			doctor: this.doctor_control ? this.doctor_control.get_value() : '',
			lead_source: this.lead_source_control ? this.lead_source_control.get_value() : '',
		};
	}

	// ── Data fetch ───────────────────────────────────────────────────────

	fetch_data() {
		const filters = this.get_filters();
		frappe.call({
			method: 'clinic.clinic.page.clinic_dashboard.clinic_dashboard.get_dashboard_data',
			args: filters,
			freeze: true,
			freeze_message: __('Loading dashboard…'),
			callback: (r) => {
				if (r && r.message) {
					this.render_all(r.message);
				}
			}
		});
	}

	// ── Render orchestrator ──────────────────────────────────────────────

	render_all(data) {
		this.render_kpis(data);
		this.render_referral_cards(data.referral_stats || {});
		this.render_status_chart(data.appointments_by_status || []);
		this.render_clinic_revenue_chart(data.revenue_by_clinic || []);
		this.render_lead_source_chart(data.lead_source_performance || []);
		this.render_top_doctors_chart(data.doctor_financial || []);
		this.render_doctor_table(data.doctor_financial || []);
	}

	// ── KPI cards ────────────────────────────────────────────────────────

	render_kpis(data) {
		const summary = data.appointment_summary || {};
		const rev = data.revenue_summary || {};
		const ref = data.referral_stats || {};
		const doctors = data.doctor_financial || [];

		// compute average per appointment
		let totalPaid = doctors.reduce((s, d) => s + (d.total_paid || 0), 0);
		let totalAppt = doctors.reduce((s, d) => s + (d.appointment_count || 0), 0);
		let avg = totalAppt > 0 ? (totalPaid / totalAppt) : 0;

		const kpis = [
			{ label: __('Total Appointments'), value: this.fmt_number(summary.total_appointments || 0), cls: 'blue' },
			{
				label: __('Total Revenue'), value: this.fmt_currency(rev.total_paid || 0), cls: 'green',
				sub: __('Invoiced: ') + this.fmt_currency(rev.total_invoiced || 0)
			},
			{ label: __('Outstanding'), value: this.fmt_currency(rev.total_outstanding || 0), cls: 'red' },
			{
				label: __('Referrals'), value: this.fmt_number(ref.total_referrals || 0), cls: 'purple',
				sub: __('Billed: ') + this.fmt_number(ref.billed_referrals || 0)
			},
			{ label: __('Avg / Appointment'), value: this.fmt_currency(avg), cls: 'amber' },
		];

		let html = kpis.map(k => `
            <div class="kpi-card ${k.cls}">
                <div class="kpi-label">${k.label}</div>
                <div class="kpi-value">${k.value}</div>
                ${k.sub ? `<div class="kpi-sub">${k.sub}</div>` : ''}
            </div>
        `).join('');

		this.$parent.find('#kpi-grid').html(html);
	}

	// ── Referral cards ───────────────────────────────────────────────────

	render_referral_cards(ref) {
		const cards = [
			{ label: __('Total Referrals'), value: ref.total_referrals || 0 },
			{ label: __('Billed Referrals'), value: ref.billed_referrals || 0 },
			{ label: __('Non-Billed'), value: ref.non_billed_referrals || 0 },
		];
		let html = cards.map(c => `
            <div class="ref-card">
                <div class="ref-value">${this.fmt_number(c.value)}</div>
                <div class="ref-label">${c.label}</div>
            </div>
        `).join('');
		this.$parent.find('#referral-grid').html(html);
	}

	// ── Charts ───────────────────────────────────────────────────────────

	_make_chart(key, el, opts) {
		// Destroy previous chart instance to free memory
		if (this.charts[key] && this.charts[key].destroy) {
			this.charts[key].destroy();
		}
		el.innerHTML = '';
		this.charts[key] = new frappe.Chart(el, opts);
	}

	render_status_chart(rows) {
		const el = this.$parent.find('#chart-status').get(0);
		if (!el) return;
		if (!rows.length) { el.innerHTML = this.empty_state(); return; }

		const colors = {
			'Scheduled': '#3b82f6', 'Waiting': '#f59e0b',
			'waiting attend': '#06b6d4', 'Attended': '#10b981',
			'Under Treatment': '#8b5cf6', 'To Bill': '#f97316',
			'Billed': '#22c55e', 'Partial Billed': '#eab308',
			'Retouch': '#6366f1',
		};

		this._make_chart('status', el, {
			type: 'donut',
			data: {
				labels: rows.map(r => r.status),
				datasets: [{ values: rows.map(r => r.count) }]
			},
			colors: rows.map(r => colors[r.status] || '#94a3b8'),
			height: 280,
			maxSlices: 10,
		});
	}

	render_clinic_revenue_chart(rows) {
		const el = this.$parent.find('#chart-clinic-revenue').get(0);
		if (!el) return;
		if (!rows.length) { el.innerHTML = this.empty_state(); return; }

		this._make_chart('clinic_rev', el, {
			type: 'bar',
			data: {
				labels: rows.map(r => r.clinic || __('Unknown')),
				datasets: [
					{ name: __('Invoiced'), values: rows.map(r => r.total_invoiced) },
					{ name: __('Paid'), values: rows.map(r => r.total_paid) },
				]
			},
			colors: ['#3b82f6', '#10b981'],
			height: 280,
			barOptions: { spaceRatio: 0.4 },
		});
	}

	render_lead_source_chart(rows) {
		const el = this.$parent.find('#chart-lead-source').get(0);
		if (!el) return;
		if (!rows.length) { el.innerHTML = this.empty_state(); return; }

		this._make_chart('lead_src', el, {
			type: 'bar',
			data: {
				labels: rows.map(r => r.lead_source || __('Unknown')),
				datasets: [
					{ name: __('New Billing'), values: rows.map(r => r.new_billing) },
					{ name: __('Repeat Billing'), values: rows.map(r => r.repeat_billing) },
				]
			},
			colors: ['#8b5cf6', '#14b8a6'],
			height: 280,
			barOptions: { spaceRatio: 0.4 },
		});
	}

	render_top_doctors_chart(rows) {
		const el = this.$parent.find('#chart-top-doctors').get(0);
		if (!el) return;
		const top = rows.slice(0, 10);
		if (!top.length) { el.innerHTML = this.empty_state(); return; }

		this._make_chart('top_docs', el, {
			type: 'bar',
			data: {
				labels: top.map(r => r.doctor_name || r.doctor),
				datasets: [
					{ name: __('Paid'), values: top.map(r => r.total_paid) },
					{ name: __('Outstanding'), values: top.map(r => r.total_outstanding) },
				]
			},
			colors: ['#10b981', '#ef4444'],
			height: 280,
			barOptions: { spaceRatio: 0.4 },
		});
	}

	// ── Doctor table ─────────────────────────────────────────────────────

	render_doctor_table(rows) {
		const tbody = this.$parent.find('#doctor-table tbody');
		if (!rows.length) {
			tbody.html(`<tr><td colspan="6" style="text-align:center;padding:30px;color:#9ca3af">${__('No data')}</td></tr>`);
			return;
		}
		let html = rows.map(r => `
            <tr>
                <td>${r.doctor_name || r.doctor}</td>
                <td>${r.appointment_count}</td>
                <td class="currency">${this.fmt_currency(r.total_invoiced)}</td>
                <td class="currency positive">${this.fmt_currency(r.total_paid)}</td>
                <td class="currency negative">${this.fmt_currency(r.total_outstanding)}</td>
                <td class="currency">${this.fmt_currency(r.average_per_appointment)}</td>
            </tr>
        `).join('');
		tbody.html(html);
	}

	// ── Helpers ──────────────────────────────────────────────────────────

	fmt_currency(v) {
		return format_currency(v, frappe.defaults.get_default('currency') || 'SAR');
	}

	fmt_number(v) {
		return (v || 0).toLocaleString();
	}

	empty_state() {
		return `<div style="text-align:center;padding:40px 0;color:#9ca3af;font-size:13px">${__('No data for selected filters')}</div>`;
	}
}